onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -568,6 +568,139 @@ void DecoderAttentionTypeAndShapeInference(ONNX_NAMESPACE::InferenceContext& ctx
+    if (ctx.getNumOutputs() > 2) {
+      ONNX_NAMESPACE::propagateElemTypeFromInputToOutput(ctx, 5, 2);
+    }
+  }
+
+  // Shape inference
+  // input 0 (input_ids) shape: (batch_size, sequence_length)
+  // output 0 (sequences) shape: (batch_size, num_return_sequences, max_length)
+  // output 1 (sequences_scores) shape: (batch_size, num_return_sequences)
+  // output 2 (scores) shape: (max_length - sequence_length, batch_size, num_beams, vocab_size)
+  if (!hasInputShape(ctx, 0)) {
+    return;
+  }
+  auto& input_ids_shape = getInputShape(ctx, 0);
+  auto& input_ids_dims = input_ids_shape.dim();
+  if (input_ids_dims.size() != 2) {
+    fail_shape_inference("Inputs 0 shall be 2 dimensions");
+  }
+  if (!(input_ids_dims[0].has_dim_value() && input_ids_dims[1].has_dim_value())) {
+    return;
+  }
+  
+  int64_t batch_size = input_ids_dims[0].dim_value();
+  int64_t sequence_length = input_ids_dims[1].dim_value();
+
+  const auto max_length = ctx.getInputData(1);
+  const auto num_beams = ctx.getInputData(3);
+  const auto num_return_sequences = ctx.getInputData(4);
+  if (num_beams == nullptr || max_length == nullptr || num_return_sequences == nullptr) {  // not initializer
+    return;
+  }
+
+  int max_length_value = 0;
+  if (!ParseScalar(max_length, max_length_value) || max_length_value <= 0) {
+    fail_shape_inference("Failed to parse max_length or it is not positive integer scalar");
+  }
+
+  int num_beams_value = 0;
+  if (!ParseScalar(num_beams, num_beams_value) || num_beams_value <= 0) {
+    fail_shape_inference("Failed to parse num_beams or it is not positive integer scalar");
+  }
+
+  int num_return_sequences_value = 0;
+  if (!ParseScalar(num_return_sequences, num_return_sequences_value) || num_return_sequences_value <= 0) {
+    fail_shape_inference("Failed to parse num_return_sequences or it is not positive integer scalar");
+  }
+
+  ONNX_NAMESPACE::TensorShapeProto sequences_shape;
+  sequences_shape.add_dim()->set_dim_value(batch_size);
+  sequences_shape.add_dim()->set_dim_value(num_return_sequences_value);
+  sequences_shape.add_dim()->set_dim_value(max_length_value);
+  updateOutputShape(ctx, 0, sequences_shape);
+
+  if (ctx.getNumOutputs() > 1) {
+    ONNX_NAMESPACE::TensorShapeProto sequences_scores_shape;
+    sequences_shape.add_dim()->set_dim_value(batch_size);
+    sequences_shape.add_dim()->set_dim_value(num_return_sequences_value);
+    updateOutputShape(ctx, 1, sequences_scores_shape);
+
+    if (ctx.getNumOutputs() > 2) {
+      ONNX_NAMESPACE::TensorShapeProto scores_shape;
+      scores_shape.add_dim()->set_dim_value(max_length_value - sequence_length);
+      scores_shape.add_dim()->set_dim_value(batch_size);
+      scores_shape.add_dim()->set_dim_value(num_beams_value);
+      scores_shape.add_dim();  // vocab_size is unknown
+      updateOutputShape(ctx, 2, scores_shape);
+    }
+  }
+}
+
+void RegisterTextGenerationSchemas() {
+  ONNX_CONTRIB_OPERATOR_SCHEMA(BeamSearch)
```

Comment:
Could you add README for beam search detailing out implementation? I see this is a big change and due to the number of operators involved and two levels of graph, this would help in tracking and extending the work easily.

Reply:
For instance, we can add for what vocab mask should (shouldn't) be used.

Reply:
There is some description of each input and each attribute in operator definition. Is that good enough?

Reply:
Lets discuss on this a bit more offline, we need to have more documentation of our features and coverage.